src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java

```diff
@@ -0,0 +1,157 @@
+import org.sqlite.util.OSInfo;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+
+public class SqliteJdbcFeature implements Feature {
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess a) {
+        RuntimeClassInitialization.initializeAtBuildTime(SQLiteJDBCLoader.VersionHolder.class);
+        RuntimeClassInitialization.initializeAtBuildTime(JDBC3DatabaseMetaData.class);
+        RuntimeClassInitialization.initializeAtBuildTime(OSInfo.class);
+        RuntimeClassInitialization.initializeAtBuildTime(LibraryLoaderUtil.class);
+        a.registerReachabilityHandler(
+                this::nativeDbReachable, method(SQLiteJDBCLoader.class, "initialize"));
+    }
+
+    private void nativeDbReachable(DuringAnalysisAccess a) {
+        handleLibraryResources();
+        registerJNICalls();
+    }
+
+    private void handleLibraryResources() {
+        String libraryPath = LibraryLoaderUtil.getNativeLibResourcePath();
+        String libraryName = LibraryLoaderUtil.getNativeLibName();
+        // Sanity check
+        if (!LibraryLoaderUtil.hasNativeLib(libraryPath, libraryName)) {
+            throw new SqliteJdbcFeatureException(
+                    "Unable to locate the required native resources for native-image. Please contact the maintainers of sqlite-jdbc.");
+        }
+
+        // libraryResource always has a leading '/'
+        String libraryResource = libraryPath + "/" + libraryName;
+        String exportLocation = System.getProperty("org.sqlite.lib.exportPath", "");
+        if (exportLocation.isEmpty()) {
+            // Do not export the library and include it in native-image instead
+            RuntimeResourceAccess.addResource(
+                    SQLiteJDBCLoader.class.getModule(), libraryResource.substring(1));
+        } else {
+            // export the required library to the specified path,
+            // the user is responsible to make sure the created native-image can actually find it.
+            Path targetPath = Paths.get(exportLocation, libraryName);
+            try (InputStream in = SQLiteJDBCLoader.class.getResourceAsStream(libraryResource)) {
+                Files.copy(in, targetPath, StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e) {
+                throw new SqliteJdbcFeatureException(e);
+            }
+        }
+    }
+
+    private void registerJNICalls() {
+        // NativeDB and DB JNI calls
+        RuntimeJNIAccess.register(NativeDB.class);
+        RuntimeJNIAccess.register(
+                fields(
+                        NativeDB.class,
+                        "pointer",
+                        "busyHandler",
+                        "commitListener",
+                        "updateListener",
+                        "progressHandler"));
```

Comment:
do those references have to be strings ? i am not familiar with GraalVM Feature.

Reply:
The references passed to `RuntimeJNIAccess#register` need to be either `Class` references, or reflection elements like `Field`, `Method` and `Constructor`. So to look them up they unfortunately need to be string references indeed.

Reply:
can't we use reflection ?

Reply:
It is using reflection. It's just using helper `fields` and `method` methods to hide error handling and to make things a little bit easier to read. I took some inspiration from how internal GraalVM Feature implementations are organised: https://github.com/oracle/graal/blob/ff5373e4db4ab37ab7c6b3e8fc8a2dc346cadce1/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JNIRegistrationUtil.java#L90-L105

I can just inline the calls if you prefer that.

Reply:
the problem i have with that is that there's strings, i would much prefer that the method references be there if we can, so it would be refactoring-proof.

Reply:
I don't understand what you're hinting at. They need to be `Method` and `Field` objects, and to my knowledge there is no other way to get those than a string-lookup.
Any refactoring that requires a change here, will also require a change in the native JNI part of the code. Those also use string lookups and would fail in much the same way if they were not updated.
The tests running on native-image, and straight up bailing out of the native-image compilation by throwing exceptions if one of the reflective lookups fail should be sufficient to catch any potential issue that can occur by forgetting to update these string references.

Reply:
I have no experience with Graalvm, that is why my questions may be naive. If there's no other way that's fine.